### PR TITLE
Metrics prototype

### PR DESCRIPTION
**Pull request by nieznanysprawiciel**
resolves: #533
resolves: #671

### CODE REVIEW

**Review by tworec — COMMENTED** [review]
This is an awesome thing. Keep going!

File: core/market/decentralized/src/matcher/cyclic.rs [hunk 1]
```diff
@@ -73,6 +84,11 @@ pub(super) async fn bcast_unsubscribes(matcher: Matcher) {
                 .discovery
                 .bcast_unsubscribes(our_and_random_ids)
                 .await?;
+
+            let end = Instant::now();
+            counter!("market.offers.unsubscribes.broadcasts", 1);
+            timing!("market.offers.unsubscribes.broadcasts.time", start, end);
```
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/cyclic.rs:90]: I don't know how could this work. You need two macros whatsoever, so i prefer to leave it like this. Or maybe you can propose something better?

**Review by jiivan — DISMISSED** [review]

**Review by Wiezzel — APPROVED** [review]
I would advise to make the `counter!` macro variadic to skip the second argument if we want to increase by 1 (this seems to be the majority of cases):
`counter!("foo")` == `counter!("foo", 1)`

**Comment by nieznanysprawiciel** [comment]
> I would advise to make the `counter!` macro variadic to skip the second argument if we want to increase by 1 (this seems to be the majority of cases):
> `counter!("foo")` == `counter!("foo", 1)`
Note that `counter!` is macro from metrics crate, not ours.

**Comment by Wiezzel** [comment]
> Note that `counter!` is macro from metrics crate, not ours.

Didn't see that. Only reviewed the payments part. I guess in this case we'll have to stick with `counter!("foo", 1)`.

**Comment by nieznanysprawiciel** [comment]
Activity exposes metric for usage vector. Interpretation of elements in this vector belongs to Grafana.
Note: that order of elements depends on Agreement and this depends from Offer created by Provider.

**Review by prekucki — APPROVED** [review]

**Review by tworec — APPROVED** [review]

**Review by jiivan — DISMISSED** [review]

**Review by mfranciszkiewicz — APPROVED** [review]

**Review by tworec — DISMISSED** [review]

**Review by jiivan — DISMISSED** [review]

File: core/payment/src/processor.rs [hunk 2]
```diff
@@ -245,6 +246,7 @@ impl PaymentProcessor {
 
         // TODO: Implement re-sending mechanism in case SendPayment fails
 
+        counter!("payment.invoices.requestor.paid", 1);
```
Comment (Wiezzel) [inline core/payment/src/processor.rs:249]: This is not semantically accurate. Single payment can be used to pay off multiple invoices if they have the same payee address. (We don't do that now, but data model allows it.) So it should be rather 'sent payments' not 'paid invoices'. If you really want paid invoices then you have to look at the length of `agreement_payments` vector here.

Comment (nieznanysprawiciel) [reply core/payment/src/processor.rs:249]: Thanks for noticing this.

**Review by tworec — APPROVED** [review]

**Pull request merged**